graphbolt/src/cuda/neighbor_sampler.cu

```diff
@@ -123,24 +124,107 @@ struct SegmentEndFunc {
+              thrust::for_each(
+                  exec_policy, iota, iota + num_rows,
+                  EdgeTypeSearch<indptr_t, etype_t>{
+                      sub_indptr.data_ptr<indptr_t>(),
+                      sliced_indptr.data_ptr<indptr_t>(),
+                      etypes.data_ptr<etype_t>(), num_fanouts, num_rows,
+                      new_sub_indptr.data_ptr<indptr_t>(),
+                      new_sliced_indptr.data_ptr<indptr_t>()});
+            }));
+        size_t tmp_storage_size = 0;
+        cub::DeviceAdjacentDifference::SubtractLeftCopy(
+            nullptr, tmp_storage_size, new_sub_indptr.data_ptr<indptr_t>(),
+            new_indegree.data_ptr<indptr_t>(), num_rows + 1, cub::Difference{},
+            stream);
+        auto tmp_storage = allocator.AllocateStorage<char>(tmp_storage_size);
+        cub::DeviceAdjacentDifference::SubtractLeftCopy(
+            tmp_storage.get(), tmp_storage_size,
+            new_sub_indptr.data_ptr<indptr_t>(),
+            new_indegree.data_ptr<indptr_t>(), num_rows + 1, cub::Difference{},
+            stream);
+      }));
+  // Discard the first element of the SubtractLeftCopy result.
+  new_indegree = new_indegree.slice(0, 1, num_rows + 2);
+  return {new_sub_indptr, new_indegree, new_sliced_indptr};
+}
+
 c10::intrusive_ptr<sampling::FusedSampledSubgraph> SampleNeighbors(
     torch::Tensor indptr, torch::Tensor indices, torch::Tensor nodes,
     const std::vector<int64_t>& fanouts, bool replace, bool layer,
     bool return_eids, torch::optional<torch::Tensor> type_per_edge,
     torch::optional<torch::Tensor> probs_or_mask) {
-  TORCH_CHECK(
-      fanouts.size() == 1, "Heterogenous sampling is not supported yet!");
   TORCH_CHECK(!replace, "Sampling with replacement is not supported yet!");
   // Assume that indptr, indices, nodes, type_per_edge and probs_or_mask
   // are all resident on the GPU. If not, it is better to first extract them
   // before calling this function.
   auto allocator = cuda::GetAllocator();
   const auto stream = cuda::GetCurrentStream();
-  const auto num_rows = nodes.size(0);
-  const auto fanout =
-      fanouts[0] >= 0 ? fanouts[0] : std::numeric_limits<int64_t>::max();
+  auto num_rows = nodes.size(0);
+  auto fanouts_pinned = torch::empty(
+      fanouts.size(),
+      c10::TensorOptions().dtype(torch::kLong).pinned_memory(true));
+  auto fanouts_pinned_ptr = fanouts_pinned.data_ptr<int64_t>();
+  for (size_t i = 0; i < fanouts.size(); i++) {
+    fanouts_pinned_ptr[i] =
+        fanouts[i] >= 0 ? fanouts[i] : std::numeric_limits<int64_t>::max();
+  }
+  // Finally, copy the adjusted fanout values to the device memory.
+  auto fanouts_device = allocator.AllocateStorage<int64_t>(fanouts.size());
+  CUDA_CALL(cudaMemcpyAsync(
+      fanouts_device.get(), fanouts_pinned_ptr,
+      sizeof(int64_t) * fanouts.size(), cudaMemcpyHostToDevice, stream));
   auto in_degree_and_sliced_indptr = SliceCSCIndptr(indptr, nodes);
```

Comment:
`auto [in_degree, sliced_indptr]`?

Reply:
When I do that, the variables can not be captured by the dispatch lambdas unfortunately.

Reply:
it seems a bug in c++ lambda?

Reply:
I checked and they say that when you unpack variables with `auto [in_degree, sliced_indptr]`, they are not proper variables. That is why they can not be captured. Probably a bug in the specification that needs to be fixed.